### PR TITLE
Terraform BigQuery Table Hive partitioning support 

### PR DESCRIPTION
**Pull request by ffung**
Fixes: https://github.com/terraform-providers/terraform-provider-google/issues/5664
As of March 2, range partioning / hive partitioning is GA, see https://cloud.google.com/bigquery/docs/release-notes. 
Note: Doesn't support `require_partition_filter` attribute as this isn't available from the used BigQuery SDK.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
bigquery: Added support for `google_bigquery_table` `hive_partitioning_options`
```

```release-note:enhancement
bigquery: Added `google_bigquery_table` `range_partitioning` to GA
```

### CODE REVIEW

**Comment by ffung** [comment]
```
± env |grep GOOG
GOOGLE_REGION=us-central1
GOOGLE_ZONE=us-central1-a

± make testacc TEST=./google TESTARGS='-run=TestAccBigQueryTable_Hive'

==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccBigQueryTable_Hive -timeout 240m -ldflags="-X=github.com/terraform-providers/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccBigQueryTable_HivePartitioning
=== PAUSE TestAccBigQueryTable_HivePartitioning
=== CONT  TestAccBigQueryTable_HivePartitioning
--- PASS: TestAccBigQueryTable_HivePartitioning (8.96s)
PASS
ok  	github.com/terraform-providers/terraform-provider-google/google	(cached)
```

File: third_party/terraform/website/docs/r/bigquery_table.html.markdown [hunk 1]
```diff
@@ -207,6 +212,23 @@ The `google_sheets_options` block supports:
     that BigQuery will skip when reading the data. At least one of `range` or
     `skip_leading_rows` must be set.
 
+The `hive_partitioning_options` block supports:
+
+* `mode` (Optional) - When set, what mode of hive partitioning to use when
```
Comment (slevenick) [inline third_party/terraform/website/docs/r/bigquery_table.html.markdown:217]: These don't look great in the downstream markdown: https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3335-old..auto-pr-3335?short_path=aa2f4fe#diff-aa2f4fe690d70a1d1d5a0f4bd663d428

Can we either add new lines before the `*` character, which should cause it to format as a list item, or quote `AUTO` and `STRINGS` in back ticks for formatting

File: third_party/terraform/website/docs/r/bigquery_table.html.markdown [hunk 2]
```diff
@@ -207,6 +212,23 @@ The `google_sheets_options` block supports:
...
+    reading data. Two modes are supported. * AUTO: automatically infer
+    partition key name(s) and type(s). * STRINGS: automatically infer
+    partition key name(s). All types are Not all storage formats support hive
+    partitioning. Requesting hive partitioning on an unsupported format will
+    lead to an error. Currently supported formats are: JSON, CSV, ORC, Avro and Parquet.
+
+* `source_uri_prefix` (Optional) - When hive partition detection is requested,
```
Comment (slevenick) [inline third_party/terraform/website/docs/r/bigquery_table.html.markdown:224]: Lets quote the `gs://...` uris here, as it should make them easier to read. Or possibly link to the guide on source_uri_prefix? This seems like a pretty complicated concept that may be better explained by the docs than a small snippet in the description

File: third_party/terraform/tests/resource_bigquery_table_test.go [hunk 3]
```diff
@@ -332,7 +353,43 @@ EOH
...
+
+resource "google_storage_bucket_object" "test" {
+	name    = "key1=20200330/init.csv"
+	content = ";"
+	bucket  = google_storage_bucket.test.name
+}
+
+resource "google_bigquery_dataset" "test" {
+		dataset_id = "%s"
+	}
```
Comment (slevenick) [inline third_party/terraform/tests/resource_bigquery_table_test.go:371]: Looks like spacing got a little mixed up in this config. The resource below is also tabbed in too much

File: third_party/terraform/website/docs/r/bigquery_table.html.markdown (hunk 2 above)
Comment (slevenick) [reply third_party/terraform/website/docs/r/bigquery_table.html.markdown:224]: On second reading of this I'm not sure it makes sense. I see that the API docs have the same text, but I would prefer if we could clean it up.

Maybe something like this guide shows: https://cloud.google.com/bigquery/docs/hive-partitioned-queries-gcs#setting_hivepartitioningoptions_using_the_rest_api

"
Note that when hive_partitioning_options.mode is set to `CUSTOM`, you must encode the partition key schema within the `source_uri_prefix` by setting hive_partitioning_options.source_uri_prefix to `gs://bucket/path_to_table/{key1:TYPE1}/{key2:TYPE2}/{key3:TYPE3}`.

When hive partitioning is requested with either `AUTO` or `STRINGS` detection, the common prefix can be either of `gs://bucket/path_to_table` or `gs://bucket/path_to_table/`.
"

File: third_party/terraform/website/docs/r/bigquery_table.html.markdown (hunk 1 above)
Comment (slevenick) [reply third_party/terraform/website/docs/r/bigquery_table.html.markdown:217]: Additionally, it looks like `CUSTOM` is a valid mode here as well. It isn't explicitly called out in the docs which is strange, but I see references to it here: https://cloud.google.com/bigquery/docs/hive-partitioned-queries-gcs#setting_hivepartitioningoptions_using_the_rest_api

File: third_party/terraform/tests/resource_bigquery_table_test.go [hunk 4]
```diff
@@ -66,7 +65,30 @@ func TestAccBigQueryTable_Kms(t *testing.T) {
...
+	t.Parallel()
+	bucketName := testBucketName()
+	resourceName := "google_bigquery_table.test"
+	datasetID := fmt.Sprintf("tf_test_%s", acctest.RandString(10))
+	tableID := fmt.Sprintf("tf_test_%s", acctest.RandString(10))
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckBigQueryTableDestroy,
```
Comment (slevenick) [inline third_party/terraform/tests/resource_bigquery_table_test.go:78]:
```suggestion
		CheckDestroy: testAccCheckBigQueryTableDestroyProducer(t),
```

File: third_party/terraform/tests/resource_bigquery_table_test.go [hunk 5]
```diff
@@ -66,7 +65,30 @@ func TestAccBigQueryTable_Kms(t *testing.T) {
 	})
 }
 
-<% unless version == 'ga' -%>
+func TestAccBigQueryTable_HivePartitioning(t *testing.T) {
+	t.Parallel()
+	bucketName := testBucketName()
+	resourceName := "google_bigquery_table.test"
+	datasetID := fmt.Sprintf("tf_test_%s", acctest.RandString(10))
```
Comment (slevenick) [inline third_party/terraform/tests/resource_bigquery_table_test.go:72]:
```suggestion
	datasetID := fmt.Sprintf("tf_test_%s", randString(t, 10))
```

File: third_party/terraform/tests/resource_bigquery_table_test.go [hunk 6]
```diff
@@ -66,7 +65,30 @@ func TestAccBigQueryTable_Kms(t *testing.T) {
...
+	tableID := fmt.Sprintf("tf_test_%s", acctest.RandString(10))
```
Comment (slevenick) [inline third_party/terraform/tests/resource_bigquery_table_test.go:73]:
```suggestion
	tableID := fmt.Sprintf("tf_test_%s", randString(t, 10))
```

File: third_party/terraform/tests/resource_bigquery_table_test.go (hunk 4 above)
Comment (ffung) [reply third_party/terraform/tests/resource_bigquery_table_test.go]: fixed

File: third_party/terraform/tests/resource_bigquery_table_test.go (hunk 6 above)
Comment (ffung) [reply third_party/terraform/tests/resource_bigquery_table_test.go]: fixed

File: third_party/terraform/tests/resource_bigquery_table_test.go (hunk 5 above)
Comment (ffung) [reply third_party/terraform/tests/resource_bigquery_table_test.go]: fixed

**Comment by ffung** [comment]
@slevenick I think I resolved all the issues, is there anything I still can do for this PR?

**Comment by slevenick** [comment]
> @slevenick I think I resolved all the issues, is there anything I still can do for this PR?

Ah, sorry lost track of this. Looks like there are conflicts now with master, could you rebase on master and I can run tests and get this merged?

**Comment by ffung** [comment]
@slevenick done.I  cannot see why the build fails, but let me know if I can help address the issue.

**Comment by slevenick** [comment]
> @slevenick done.I cannot see why the build fails, but let me know if I can help address the issue.

Looks like the files aren't formatted correctly:

```
gofmt needs running on the following files:
./google/resource_bigquery_table_test.go
./google/resource_bigquery_table.go```

You can generate the downstream files, run `gofmt` on them and then copy them back into Magic Modules to fix this

**Comment by ffung** [comment]
@slevenick ran gofmt, build still fails though

**Comment by slevenick** [comment]
> @slevenick ran gofmt, build still fails though

Looks like the problem is now:
```
google/resource_bigquery_table_test.go:69:30: not enough arguments in call to testBucketName
	have ()
	want (*testing.T)
```

You can run `make test` in the generated provider to make sure everything passes

**Comment by ffung** [comment]
@slevenick ah sorry that was sloppy, tests are running succesfully now locally, but the builds still seems to fail.

**Comment by slevenick** [comment]
Sorry the process is a little rough around the edges. Looks like we enforce that gofmt must be run with a specific flag 😞 

```
Step #22 - "tpgb-test": google-beta/resource_bigquery_table.go:336: File is not `gofmt`-ed with `-s` (gofmt)
Step #22 - "tpgb-test": 			"range_partitioning": &schema.Schema{```

**Comment by ffung** [comment]
@slevenick _that_ should fix it;-)

**Comment by LoekL** [comment]
FYI I tried using this today, no dice. The table does get created:

Screenshot 2020-06-23 18 11 31

But there's no sign of the partitions in the table schema:

Screenshot 2020-06-23 18 12 31

If I however run:

```sh
bq mkdef \
  --autodetect \
  --ignore_unknown_values \
  --source_format=NEWLINE_DELIMITED_JSON \
  --hive_partitioning_mode=CUSTOM \
  --hive_partitioning_source_uri_prefix=gs://anhistous-metonymic-7578834-analytics-staging-batched/data_type=jsonl/event_schema=playfab/{event_category:STRING}/{event_environment:STRING}/{event_date:DATE}/{event_hour:STRING}/{event_minute:STRING} \
  gs://anhistous-metonymic-7578834-analytics-staging-batched/data_type=jsonl/event_schema=playfab/\* \
  AnalyticsEnvironment:STRING,PlayFabEnvironment:STRING,SourceType:STRING,Source:STRING,EventNamespace:STRING,TitleId:STRING,GroupBatchId:STRING,BatchId:STRING,EventId:STRING,EventName:STRING,EntityType:STRING,EntityId:STRING,Timestamp:TIMESTAMP,ReceivedTimestamp:TIMESTAMP,BatchedTimestamp:TIMESTAMP,BatchJobName:STRING,EventAttributes:STRING \
  > /Users/loek/Desktop/events_playfab_staging_hive_partitioned_batched
```

Followed by:
```sh
bq mk --table --location=US --external_table_definition=/Users/loek/Desktop/events_playfab_staging_hive_partitioned_batched external.events_playfab_hive_partitioned_batched
```

It does work:

Screenshot 2020-06-23 18 15 54

&

Screenshot 2020-06-23 18 16 00

The only difference within the BQ UI I can see is that when going via the `bq` CLI, I don't see `Compression = GZIP` in the table details. Am using `Terraform v0.12.18` & `provider.google v3.26.0`.

**Comment by slevenick** [comment]
@LoekL can you create a new issue on https://github.com/terraform-providers/terraform-provider-google with steps to reproduce this problem so that we can look into this?

**Comment by ffung** [comment]
yes @LoekL please raise an issue, I cannot reproduce your problem. If I apply this, hive partitioning works just fine.

```
resource "google_bigquery_dataset" "test" {
  dataset_id                  = "foo"
  friendly_name               = "test"
  description                 = "This is a test description"
  location                    = "europe-west3"
  default_table_expiration_ms = 3600000

  labels = {
    env = "default"
  }
}

resource "google_bigquery_table" "test" {
  dataset_id = google_bigquery_dataset.test.dataset_id
  table_id   = "bar"

  external_data_configuration {
    source_format = "PARQUET"
    autodetect = true
    source_uris= ["gs://project-2resdfdsafsdf43/test/*"]

    hive_partitioning_options {
      mode = "CUSTOM"
      source_uri_prefix = "gs://project-2resdfdsafsdf43/test/{part:STRING}/{dt:DATE}"
    }
  }
}
```

**Comment by LoekL** [comment]
@slevenick @ffung - see: https://github.com/terraform-providers/terraform-provider-google/issues/6693

**Pull request merged**